mlflow/utils/file_utils.py

```diff
@@ -588,6 +597,107 @@ def download_file_using_http_uri(http_uri, download_path, chunk_size=100000000,
+    # and write chunks without overwriting.
+    open(download_path, "w").close()
+    starting_index = 0
+    if uri_type == ArtifactCredentialType.GCP_SIGNED_URL or uri_type is None:
+        # GCP files could be transcoded, in which case the range header is ignored.
+        # Test if this is the case by downloading one chunk and seeing if it's larger than the
+        # requested size. If yes, let that be the file; if not, continue downloading more chunks.
+        download_chunk(
+            range_start=0,
+            range_end=chunk_size - 1,
+            headers=headers,
+            download_path=download_path,
+            http_uri=http_uri,
+        )
+        downloaded_size = os.path.getsize(download_path)
+        # If downloaded size was equal to the chunk size it would have been downloaded serially,
+        # so we don't need to consider this here
+        if downloaded_size > chunk_size:
+            return {}
+        else:
+            starting_index = 1
+
+    failed_downloads = {}
+
+    def run_download(range_start, range_end):
+        with tempfile.TemporaryDirectory() as tmpdir:
+            temp_file = os.path.join(tmpdir, "error_messages.txt")
+            download_proc = _exec_cmd(
+                cmd=[
+                    sys.executable,
+                    download_cloud_file_chunk.__file__,
+                    "--range-start",
+                    range_start,
+                    "--range-end",
+                    range_end,
+                    "--headers",
+                    json.dumps(headers or {}),
+                    "--download-path",
+                    download_path,
+                    "--http-uri",
+                    http_uri,
+                    "--temp-file",
+                    temp_file,
+                ],
+                throw_on_error=True,
+                synchronous=False,
+                capture_output=True,
+                stream_output=False,
+                env=env,
+            )
+            _, stderr = download_proc.communicate()
+            if download_proc.returncode != 0:
+                with open(temp_file, "r") as f:
+                    file_contents = f.read()
+                    if file_contents:
+                        return json.loads(file_contents)
+                    else:
+                        raise Exception(
+                            "Error from download_cloud_file_chunk not captured, "
+                            f"return code {download_proc.returncode}, stderr {stderr}"
+                        )
+
+    with ThreadPoolExecutor(max_workers=MAX_PARALLEL_DOWNLOAD_WORKERS) as p:
+        futures = {}
+        for i in range(starting_index, num_requests):
+            range_start = i * chunk_size
+            range_end = range_start + chunk_size - 1
+            futures[p.submit(run_download, range_start, range_end)] = i
+
+        for future in as_completed(futures):
```

Comment:
```suggestion
        failed_downloads = {}
        for future in as_completed(futures):
```

Can we move `failed_downloads` here?